File: typescript/src/stable/stablePool.ts
import { MAX_UINT256, MAX_BALANCE } from '../constants';
import { MathSol } from '../utils/math';
import { toRawUndoRateRoundDown } from '../vault/utils';
import {
    MaxSingleTokenRemoveParams,
    MaxSwapParams,
    type PoolBase,
    Rounding,
    SwapKind,
    type SwapParams,
} from '../vault/types';
import { StableMutable } from './data';
import {
    _computeOutGivenExactIn,
    _computeInGivenExactOut,
    _computeInvariant,
    _computeBalance,
} from './stableMath';

export class Stable implements PoolBase {
    public amp: bigint;

    constructor(poolState: StableMutable) {
        this.amp = poolState.amp;
    }

    /**
     * Returns the max amount that can be swapped in relation to the swapKind.
     * @param maxSwapParams
     * @returns GivenIn: Returns the max amount in. GivenOut: Returns the max amount out.
     */
    getMaxSwapAmount(maxSwapParams: MaxSwapParams): bigint {
        const {
            balancesLiveScaled18,
            indexIn,
            indexOut,
            tokenRates,
            scalingFactors,
            swapKind,
        } = maxSwapParams;
        if (swapKind === SwapKind.GivenIn) {
            // MAX_BALANCE comes from SC limit and is max pool can hold
            const diff = MAX_BALANCE - balancesLiveScaled18[indexIn];
            // Scale to token in (and remove rate)
            return toRawUndoRateRoundDown(
                diff,
                scalingFactors[indexIn],
                tokenRates[indexIn],
            );
        }
        // 99% of token out balance
        const max = MathSol.mulDownFixed(
            990000000000000000n,
            balancesLiveScaled18[indexOut],
        );
        // Scale to token out
        return toRawUndoRateRoundDown(
            max,
            scalingFactors[indexOut],
            tokenRates[indexOut],
        );
    }

    getMaxSingleTokenAddAmount(): bigint {
        return MAX_UINT256;
    }

    getMaxSingleTokenRemoveAmount(
        maxRemoveParams: MaxSingleTokenRemoveParams,
    ): bigint {
        const {
            isExactIn,
            totalSupply,
            tokenOutBalance,
            tokenOutScalingFactor,
            tokenOutRate,
        } = maxRemoveParams;
        return this.getMaxSwapAmount({
            swapKind: isExactIn ? SwapKind.GivenIn : SwapKind.GivenOut,
            balancesLiveScaled18: [totalSupply, tokenOutBalance],
            tokenRates: [1000000000000000000n, tokenOutRate],
            scalingFactors: [1000000000000000000n, tokenOutScalingFactor],
            indexIn: 0,
            indexOut: 1,
        });
    }

    onSwap(swapParams: SwapParams): bigint {
        const {
            swapKind,
            balancesLiveScaled18: balancesScaled18,
            indexIn,
            indexOut,
            amountGivenScaled18,
        } = swapParams;
        const invariant = _computeInvariant(this.amp, balancesScaled18);

        if (swapKind === SwapKind.GivenIn) {
            return _computeOutGivenExactIn(
                this.amp,
                balancesScaled18,
                indexIn,
                indexOut,
                amountGivenScaled18,
                invariant,
            );
        }
        return _computeInGivenExactOut(
            this.amp,
            balancesScaled18,
            indexIn,
            indexOut,
            amountGivenScaled18,
            invariant,
        );
    }
    computeInvariant(
        balancesLiveScaled18: bigint[],
        rounding: Rounding,
    ): bigint {
        let invariant = _computeInvariant(this.amp, balancesLiveScaled18);
        if (invariant > 0) {
            invariant =
                rounding == Rounding.ROUND_DOWN ? invariant : invariant + 1n;
        }
        return invariant;
    }
    computeBalance(
        balancesLiveScaled18: bigint[],
        tokenInIndex: number,
        invariantRatio: bigint,
    ): bigint {
        return _computeBalance(
            this.amp,
            balancesLiveScaled18,
            MathSol.mulUpFixed(
                this.computeInvariant(balancesLiveScaled18, Rounding.ROUND_UP),
                invariantRatio,
            ),
            tokenInIndex,
        );
    }
}

File: rust/src/pools/stable/stable_pool.rs
use crate::common::errors::PoolError;
use crate::common::maths::mul_down_fixed;
use crate::common::pool_base::PoolBase;
use crate::common::types::{Rounding, SwapKind, SwapParams};
use crate::pools::stable::stable_data::StableMutable;
use crate::pools::stable::stable_math::{
    compute_balance, compute_in_given_exact_out, compute_invariant, compute_out_given_exact_in,
    _MAX_INVARIANT_RATIO, _MIN_INVARIANT_RATIO,
};
use alloy_primitives::U256;

/// Stable pool implementation
pub struct StablePool {
    pub amp: U256,
}

impl StablePool {
    /// Create a new stable pool
    pub fn new(pool_state: StableMutable) -> Self {
        Self {
            amp: pool_state.amp,
        }
    }
}

impl PoolBase for StablePool {
    fn get_maximum_invariant_ratio(&self) -> U256 {
        U256::from(_MAX_INVARIANT_RATIO)
    }

    fn get_minimum_invariant_ratio(&self) -> U256 {
        U256::from(_MIN_INVARIANT_RATIO)
    }

    fn on_swap(&self, swap_params: &SwapParams) -> Result<U256, PoolError> {
        let invariant = compute_invariant(&self.amp, &swap_params.balances_live_scaled_18)?;

        let result = match swap_params.swap_kind {
            SwapKind::GivenIn => compute_out_given_exact_in(
                &self.amp,
                &swap_params.balances_live_scaled_18,
                swap_params.token_in_index,
                swap_params.token_out_index,
                &swap_params.amount_scaled_18,
                &invariant,
            )?,
            SwapKind::GivenOut => compute_in_given_exact_out(
                &self.amp,
                &swap_params.balances_live_scaled_18,
                swap_params.token_in_index,
                swap_params.token_out_index,
                &swap_params.amount_scaled_18,
                &invariant,
            )?,
        };

        Ok(result)
    }

    fn compute_invariant(
        &self,
        balances_live_scaled18: &[U256],
        rounding: Rounding,
    ) -> Result<U256, PoolError> {
        let mut invariant = compute_invariant(&self.amp, balances_live_scaled18)?;

        if invariant > U256::ZERO {
            match rounding {
                Rounding::RoundDown => {}
                Rounding::RoundUp => {
                    invariant += U256::ONE;
                }
            }
        }

        Ok(invariant)
    }

    fn compute_balance(
        &self,
        balances_live_scaled18: &[U256],
        token_in_index: usize,
        invariant_ratio: &U256,
    ) -> Result<U256, PoolError> {
        let invariant = self.compute_invariant(balances_live_scaled18, Rounding::RoundUp)?;
        let scaled_invariant = mul_down_fixed(&invariant, invariant_ratio)?;

        compute_balance(
            &self.amp,
            balances_live_scaled18,
            &scaled_invariant,
            token_in_index,
        )
    }
}
